enterprise/server/cmd/action_runner/main.go

```diff
@@ -0,0 +1,662 @@
+		Children: []*bespb.BuildEventId{
+			{Id: &bespb.BuildEventId_Progress{Progress: &bespb.BuildEventId_ProgressId{OpaqueCount: count + 1}}},
+		},
+		Payload: &bespb.BuildEvent_Progress{Progress: &bespb.Progress{
+			// Only outputting to stderr for now, like Bazel does.
+			Stderr: output,
+		}},
+	})
+}
+
+// TODO: Handle shell variable expansion. Probably want to run this with sh -c
+func bazelArgs(cmd string) ([]string, error) {
+	tokens, err := shlex.Split(cmd)
+	if err != nil {
+		return nil, err
+	}
+	if tokens[0] == "bazel" || tokens[0] == "bazelisk" {
+		tokens = tokens[1:]
+	}
+	return tokens, nil
+}
+
+func cloneRepo(ctx context.Context) error {
+	if err := os.Mkdir(repoDirName, 0o775); err != nil {
+		return fmt.Errorf("mkdir %q: %s", repoDirName, err)
+	}
+	if err := os.Chdir(repoDirName); err != nil {
+		return fmt.Errorf("cd %q: %s", repoDirName, err)
+	}
+	if err := runCommand(ctx, "git", []string{"init"}, emptyEnv, os.Stderr); err != nil {
+		return err
+	}
+	authURL, err := authRepoURL()
+	if err != nil {
+		return err
+	}
+	if err := runCommand(ctx, "git", []string{"remote", "add", "origin", authURL}, emptyEnv, os.Stderr); err != nil {
+		return err
+	}
+	if err := runCommand(ctx, "git", []string{"fetch", "origin", *commitSHA}, emptyEnv, os.Stderr); err != nil {
+		return err
+	}
+	if err := runCommand(ctx, "git", []string{"checkout", *commitSHA}, emptyEnv, os.Stderr); err != nil {
+		return err
+	}
+	return nil
+}
+
+func authRepoURL() (string, error) {
+	user := os.Getenv(repoUserEnvVarName)
+	token := os.Getenv(repoTokenEnvVarName)
+	if user == "" && token == "" {
+		return *repoURL, nil
+	}
+	u, err := url.Parse(*repoURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse repo URL %q: %s", *repoURL, err)
+	}
+	u.User = url.UserPassword(user, token)
+	return u.String(), nil
+}
+
+func invocationURL(invocationID string) string {
+	urlPrefix := *besResultsURL
+	if !strings.HasSuffix(urlPrefix, "/") {
+		urlPrefix = urlPrefix + "/"
+	}
+	return urlPrefix + invocationID
+}
+
+func readConfig() (*workflowconf.BuildBuddyConfig, error) {
+	f, err := os.Open(actionsConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("open %q: %s", actionsConfigPath, err)
+	}
+	c, err := workflowconf.NewConfig(f)
+	if err != nil {
+		return nil, fmt.Errorf("read %q: %s", actionsConfigPath, err)
+	}
+	return c, nil
+}
+
+func matchesAnyTrigger(action *workflowconf.Action, event, branch string) bool {
+	if action.Triggers == nil {
+		return false
+	}
+	if pushCfg := action.Triggers.Push; pushCfg != nil && event == pushEventName {
+		return matchesAnyBranch(pushCfg.Branches, branch)
+	}
+
+	if prCfg := action.Triggers.PullRequest; prCfg != nil && event == pullRequestEventName {
+		return matchesAnyBranch(prCfg.Branches, branch)
+	}
+	return false
+}
+
+func matchesAnyBranch(branches []string, branch string) bool {
+	if branches == nil {
+		return false
+	}
```

Comment:
you can just remove this -- a nil slice is OK to iterate over.

"A nil slice has a length and capacity of 0 and has no underlying array."

Reply:
TIL again! thanks